File: 2022/Eduwelfare/app/src/main/java/com/mallup/eduwelfare/service/MyLocation.java
package com.mallup.eduwelfare.service;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.location.Location;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;

import com.mallup.eduwelfare.common.HNApplication;
import com.mallup.eduwelfare.delegator.HNCommTran;
import com.mallup.eduwelfare.delegator.HNCommTranInterface;
import com.mallup.eduwelfare.util.LogUtil;

import org.json.JSONObject;

/**
 * Created by skcrackers on 2/19/18.
 */

public class MyLocation extends Service
{
    private static final String TAG = "SeongKwon";
    private LocationManager mLocationManager = null;
    private static final int LOCATION_INTERVAL = 30000;
    private static final float LOCATION_DISTANCE = 10f;
    private HNCommTran mHNCommTran;

    private class LocationListener implements android.location.LocationListener
    {
        Location mLastLocation;

        public LocationListener(String provider)
        {
            Log.e(TAG, "LocationListener " + provider);
            mLastLocation = new Location(provider);
        }

        @Override
        public void onLocationChanged(Location location)
        {
            Log.e(TAG, "onLocationChanged: " + location);
            Log.e(TAG, "latitude: " + location.getLatitude());
            Log.e(TAG, "longitude: " + location.getLongitude());

            mLastLocation.set(location);

            new Thread(new Runnable() {
                public void run() {
                    try {
                        JSONObject jObj = new JSONObject();
                        mHNCommTran = new HNCommTran(new HNCommTranInterface() {
                            @Override
                            public void recvMsg(String tranCode, String params) {
                                LogUtil.e("recv location_update1 : " + tranCode + " : " + params);
                                if (tranCode.startsWith(HNApplication.URL + "/member/location_update.asp")) {
                                    LogUtil.e("recv location_update2 : " + tranCode + " : " + params);
                                }
                            }
                        });
                        mHNCommTran.sendMsg(HNApplication.URL + "/member/location_update.asp?latitude=" + mLastLocation.getLatitude() +
                                "&longitude=" + mLastLocation.getLongitude() + "&id=&DeviceId=" + HNApplication.mDeviceId, jObj);
                        return;
                    } catch (Exception localException) {
                        localException.printStackTrace();
                    }
                }
            }).start();
        }

        @Override
        public void onProviderDisabled(String provider)
        {
            Log.e(TAG, "onProviderDisabled: " + provider);
        }

        @Override
        public void onProviderEnabled(String provider)
        {
            Log.e(TAG, "onProviderEnabled: " + provider);
        }

        @Override
        public void onStatusChanged(String provider, int status, Bundle extras)
        {
            Log.e(TAG, "onStatusChanged: " + provider);
        }
    }

    LocationListener[] mLocationListeners = new LocationListener[] {
            new LocationListener(LocationManager.GPS_PROVIDER),
            new LocationListener(LocationManager.NETWORK_PROVIDER)
    };

    @Override
    public IBinder onBind(Intent arg0)
    {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId)
    {
        Log.e(TAG, "onStartCommand");
        super.onStartCommand(intent, flags, startId);
        return START_STICKY;
    }

    @Override
    public void onCreate()
    {
        Log.e(TAG, "onCreate");
        initializeLocationManager();
        try {
            mLocationManager.requestLocationUpdates(
                    LocationManager.NETWORK_PROVIDER, LOCATION_INTERVAL, LOCATION_DISTANCE,
                    mLocationListeners[1]);
        } catch (SecurityException ex) {
            Log.i(TAG, "fail to request location update, ignore", ex);
        } catch (IllegalArgumentException ex) {
            Log.d(TAG, "network provider does not exist, " + ex.getMessage());
        }
        try {
            mLocationManager.requestLocationUpdates(
                    LocationManager.GPS_PROVIDER, LOCATION_INTERVAL, LOCATION_DISTANCE,
                    mLocationListeners[0]);
        } catch (SecurityException ex) {
            Log.i(TAG, "fail to request location update, ignore", ex);
        } catch (IllegalArgumentException ex) {
            Log.d(TAG, "gps provider does not exist " + ex.getMessage());
        }
    }

    @Override
    public void onDestroy()
    {
        Log.e(TAG, "onDestroy");
        super.onDestroy();
        if (mLocationManager != null) {
            for (int i = 0; i < mLocationListeners.length; i++) {
                try {
                    mLocationManager.removeUpdates(mLocationListeners[i]);
                } catch (Exception ex) {
                    Log.i(TAG, "fail to remove location listners, ignore", ex);
                }
            }
        }
    }

    private void initializeLocationManager() {
        Log.e(TAG, "initializeLocationManager");
        if (mLocationManager == null) {
            mLocationManager = (LocationManager) getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
        }
    }
}

File: 2025/HotdealLive/app/src/main/java/com/creator/hotdeallive2/service/MyLocation.kt
package com.creator.hotdeallive2.service

import android.app.*
import android.content.Intent
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import com.creator.hotdeallive2.common.HNApplication
import com.creator.hotdeallive2.delegator.HNCommTran
import com.creator.hotdeallive2.delegator.HNCommTranInterface
import com.creator.hotdeallive2.util.*
import org.json.JSONObject

/**
 * Created by skcrackers on 2/19/18.
 */
class MyLocation : Service() {
    private var mLocationManager: LocationManager? = null
    private var mHNCommTran: HNCommTran? = null

    inner class LocationListener(provider: String) : android.location.LocationListener {
        var mLastLocation: Location

        init {
            Log.e(TAG, "LocationListener $provider")
            mLastLocation = Location(provider)
        }

        override fun onLocationChanged(location: Location) {
            Log.e(TAG, "onLocationChanged: $location")
            Log.e(TAG, "latitude: " + location.latitude)
            Log.e(TAG, "longitude: " + location.longitude)
            mLastLocation.set(location)
            Thread(Runnable {
                try {
                    val jObj = JSONObject()
                    mHNCommTran = HNCommTran(object : HNCommTranInterface {
                        override fun recvMsg(tranCode: String?, params: String) {
                            LogUtil.e("recv location_update1 : $tranCode : $params")
                            if (tranCode!!.startsWith(HNApplication.URL + "/member/location_update.asp")) {
                                LogUtil.e("recv location_update2 : $tranCode : $params")
                            }
                        }
                    })
                    mHNCommTran!!.sendMsg(
                        HNApplication.Companion.URL + "/member/location_update.asp?latitude=" + mLastLocation.latitude +
                                "&longitude=" + mLastLocation.longitude + "&id=&DeviceId=" + HNApplication.Companion.mDeviceId,
                        jObj
                    )
                    return@Runnable
                } catch (localException: Exception) {
                    localException.printStackTrace()
                }
            }).start()
        }

        override fun onProviderDisabled(provider: String) {
            Log.e(TAG, "onProviderDisabled: $provider")
        }

        override fun onProviderEnabled(provider: String) {
            Log.e(TAG, "onProviderEnabled: $provider")
        }

        override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {
            Log.e(TAG, "onStatusChanged: $provider")
        }
    }

    var mLocationListeners = arrayOf<LocationListener>(
        LocationListener(LocationManager.GPS_PROVIDER),
        LocationListener(LocationManager.NETWORK_PROVIDER)
    )

    override fun onBind(arg0: Intent): IBinder? {
        return null
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        Log.e(TAG, "onStartCommand")
        super.onStartCommand(intent, flags, startId)
        return START_STICKY
    }

    override fun onCreate() {
        Log.e(TAG, "onCreate")
        initializeLocationManager()
        try {
            mLocationManager!!.requestLocationUpdates(
                LocationManager.NETWORK_PROVIDER, LOCATION_INTERVAL.toLong(), LOCATION_DISTANCE,
                mLocationListeners[1]
            )
        } catch (ex: SecurityException) {
            Log.i(TAG, "fail to request location update, ignore", ex)
        } catch (ex: IllegalArgumentException) {
            Log.d(TAG, "network provider does not exist, " + ex.message)
        }
        try {
            mLocationManager!!.requestLocationUpdates(
                LocationManager.GPS_PROVIDER, LOCATION_INTERVAL.toLong(), LOCATION_DISTANCE,
                mLocationListeners[0]
            )
        } catch (ex: SecurityException) {
            Log.i(TAG, "fail to request location update, ignore", ex)
        } catch (ex: IllegalArgumentException) {
            Log.d(TAG, "gps provider does not exist " + ex.message)
        }
    }

    override fun onDestroy() {
        Log.e(TAG, "onDestroy")
        super.onDestroy()
        if (mLocationManager != null) {
            for (i in mLocationListeners.indices) {
                try {
                    mLocationManager!!.removeUpdates(mLocationListeners[i])
                } catch (ex: Exception) {
                    Log.i(TAG, "fail to remove location listners, ignore", ex)
                }
            }
        }
    }

    private fun initializeLocationManager() {
        Log.e(TAG, "initializeLocationManager")
        if (mLocationManager == null) {
            mLocationManager =
                applicationContext.getSystemService(LOCATION_SERVICE) as LocationManager
        }
    }

    companion object {
        private const val TAG = "SeongKwon"
        private const val LOCATION_INTERVAL = 30000
        private const val LOCATION_DISTANCE = 10f
    }
}